### ISSUE TITLE
Docker fails to start on Windows 10 : Cancellation token triggered before we finished reading from the stream.

### ISSUE THREAD

**Issue by devlace** [docker/for-win#10991]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 0C47A3B3-48AD-493B-A747-84479E32F222/20210421115517

### Actual behavior
Docker fails to start with error:
"System.Threading.Tasks.TaskCanceledException:
Cancellation token triggered before we finished reading from the stream."

### Expected behavior
Docker successfully starts

### Information

  - Windows Version: Windows 10 Build 19042.928
  - Docker Desktop Version: 3.3.1
  - WSL2 or Hyper-V backend?  WSL 2
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

### Steps to reproduce the behavior

I tried installing docker, starting it, but it fails to start with the error: Cancellation token triggered before we finished reading from the stream.

**Comment by devlace** [comment]
Full stack trace: 

System.Threading.Tasks.TaskCanceledException:
Cancellation token triggered before we finished reading from the stream.
   at HttpOverStream.ByLineReader.<ReadLineAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at HttpOverStream.Client.DialMessageHandler.<SendAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.Core.Logging.LoggingMessageHandler.<SendAsync>d__4.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.Core\Logging\LoggingMessageHandler.cs:line 30
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.Core.HttpClientExtensions.<GetJsonAsync>d__0`1.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.Core\HttpClientExtensions.cs:line 16
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.ApiServices.LifecycleClient.<DockerStartAsync>d__10.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.ApiServices\LifecycleClient.cs:line 122
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.LinuxkitDaemonStartup.<StartAsync>d__5.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.Desktop\LinuxkitDaemonStartup.cs:line 30
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.Engines.WSL2.LinuxWSL2Engine.<DoStartAsync>d__26.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.Desktop\Engines\WSL2\LinuxWSL2Engine.cs:line 165
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.ApiServices.StateMachines.TaskExtensions.<WrapAsyncInCancellationException>d__0.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.ApiServices\StateMachines\TaskExtensions.cs:line 20
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.ApiServices.StateMachines.StartTransition.<DoRunAsync>d__5.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.ApiServices\StateMachines\StartTransition.cs:line 29
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Docker.ApiServices.StateMachines.StartTransition.<DoRunAsync>d__5.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.ApiServices\StateMachines\StartTransition.cs:line 38
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.ApiServices.StateMachines.EngineStateMachine.<StartAsync>d__14.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.ApiServices\StateMachines\EngineStateMachine.cs:line 69
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Docker.Engines.Engines.<RestartAsync>d__26.MoveNext() in C:\workspaces\PR-15633\src\github.com\docker\pinata\win\src\Docker.Desktop\Engines\Engines.cs:line 265

**Comment by musahaidari** [comment]
I have the same problem too

**Comment by devlace** [comment]
I downgraded to 2.5.0.1 and now docker starts up fine and running.

**Comment by ivanovvlad9626** [comment]
i have the same problem too
![image](https://user-images.githubusercontent.com/29759826/115940239-75efa500-a4a9-11eb-9879-44103e852d9d.png)
![image](https://user-images.githubusercontent.com/29759826/115940255-80aa3a00-a4a9-11eb-9658-dde94ea833b6.png)

**Comment by peterfelts** [comment]
I have the same issue, as of 7/21/2021 on Windows 10 (build 19043.1110) and WSL2. I don't know if this did the trick or if this was just a coincidence, but I did the following and now Docker is running without throwing this error at startup:
1. Uninstalled Docker Desktop
2. Stopped and started the WSL2 Ubuntu distro
  - wsl --shutdown
  - wsl -d Ubuntu
3. Uninstalled and reinstalled Docker desktop (Docker version 20.10.7, build f0df350).

**Comment by CodisRedding** [comment]
Today, out of nowhere the latest version of Docker Desktop 3.5.2 one of our devs got this error and couldn't run local dev anymore. We tried everything from re-installing to all-the-things on stackoverflow but only uninstalling and installing 2.5.0.1 worked for us. Thanks @devlace

**Comment by peterfelts** [comment]
I don't know if this was caused by resource usage, but I was having other issues with Docker (i/o timeouts) and my system was in general running very sluggishly. I noticed that both my CPU and memory were pegged. I'm using WSL2 on Windows and had _thought_ that I had it configured to only use a max of 10GB of memory and only one CPU core but that didn't seem to be happening. 

I recreated my .wslconfig file and restarted WSL and now Docker is running fine as my system isn't running out of memory/CPU.

**Comment by z-kovacs** [comment]
Same issue - Docker 3.5.2  & Windows 10 (build 19043.1110) and WSL2 , had to downgrade to 2.5.0.1.

**Comment by peterfelts** [comment]
Another thing that I found that helps a lot: I was running my containers inside WSL2 but was building and debugging from the Windows environment/shell/file system. Upon advice from a teammate I switched my dev environment to run completely* within WSL2, meaning that I moved my enlistment to the WSL2 file system and I ran git, mpn, Node, etc. from within a Bash shell inside WSL2. My system is now waaaaaay more responsive and usable (this is subjective but it sure feels accurate). 

* I'm still running VSCode from the Windows environment with an extension that lets it communicate with the WSL2-hoated shell.
